Q&A Beta

nick-funk opened the pull request:
## What does this do?

Adds full Q&A features to the stream.

Figma Spec: https://www.figma.com/file/IJu2JNZFNvYiolJWEdNeY7al/Talk-v5---Stream?node-id=4211%3A0

## How to test?

- Enable Q&A using the new configure tab
- Add an expert
- Go back to the stream and ask questions as a commenter
- Respond to them as the expert user
- See how the functionality works with the 3 new tabs

nick-funk committed:
[CORL-863] Allow streams to be converted/configured to Q&A (#2809)

* Create preliminary schema changes for Q&A

Adds a mode and expert User onto the StorySettings.
Adds a mode selection drop down on a story's
Configure tab.

CORL-863

* Allow multiple experts, remove form elements from search

Makes the previous expert user on a Q&A story now
an array of users who can be assigned.

Converts the previous form based search that was
pulled from the admin community area into a set
of events built on callbacks.

CORL-863

* Create addExpertToStory mutation

CORL-863

* Create removeExpertFromStory mutation

CORL-863

* Conditionally show the the expert selection options

CORL-863

* Create a dropdown search control for Q&A experts

CORL-863

* Fixing up tests to match new QA stream options

Adds a few localization fixes to make sure tests
pass.
Updates existing snapshots.

CORL-863

* Add load more button to expert search list

CORL-863

* Update experts query to match react upgrades

CORL-863

* Move the Q&A config to its own section under stream config

Create enable and disable Q&A mutations/button toggle.

CORL-863

* Fix alignment and layout of expert list items

CORL-863

* Define translations and update tests

CORL-863

* Use official copy for Q&A config

CORL-863

nick-funk committed:
[CORL-856] Show expert badge on comments (#2829)

* Create preliminary schema changes for Q&A

Adds a mode and expert User onto the StorySettings.
Adds a mode selection drop down on a story's
Configure tab.

CORL-863

* Allow multiple experts, remove form elements from search

Makes the previous expert user on a Q&A story now
an array of users who can be assigned.

Converts the previous form based search that was
pulled from the admin community area into a set
of events built on callbacks.

CORL-863

* Create addExpertToStory mutation

CORL-863

* Create removeExpertFromStory mutation

CORL-863

* Conditionally show the the expert selection options

CORL-863

* Create a dropdown search control for Q&A experts

CORL-863

* Fixing up tests to match new QA stream options

Adds a few localization fixes to make sure tests
pass.
Updates existing snapshots.

CORL-863

* Add load more button to expert search list

CORL-863

* Update experts query to match react upgrades

CORL-863

* Move the Q&A config to its own section under stream config

Create enable and disable Q&A mutations/button toggle.

CORL-863

* Fix alignment and layout of expert list items

CORL-863

* Define translations and update tests

CORL-863

* Use official copy for Q&A config

CORL-863

* Show expert badges on comments when Q&A is enabled

CORL-856

* Update mutation responses and tests due to added expert fields

CORL-856

* Use EXPERT user tags to denote expert users

Removes the need for viewerIsExpert and
authorIsExpert loader/resolvers on Stories
and Comments respectively.

CORL-856

nick-funk committed:
[CORL-879] Add an unanswered tab to stream when in Q&A mode (#2838)

* Create preliminary schema changes for Q&A

Adds a mode and expert User onto the StorySettings.
Adds a mode selection drop down on a story's
Configure tab.

CORL-863

* Allow multiple experts, remove form elements from search

Makes the previous expert user on a Q&A story now
an array of users who can be assigned.

Converts the previous form based search that was
pulled from the admin community area into a set
of events built on callbacks.

CORL-863

* Create addExpertToStory mutation

CORL-863

* Create removeExpertFromStory mutation

CORL-863

* Conditionally show the the expert selection options

CORL-863

* Create a dropdown search control for Q&A experts

CORL-863

* Fixing up tests to match new QA stream options

Adds a few localization fixes to make sure tests
pass.
Updates existing snapshots.

CORL-863

* Add load more button to expert search list

CORL-863

* Update experts query to match react upgrades

CORL-863

* Move the Q&A config to its own section under stream config

Create enable and disable Q&A mutations/button toggle.

CORL-863

* Fix alignment and layout of expert list items

CORL-863

* Define translations and update tests

CORL-863

* Use official copy for Q&A config

CORL-863

* Show expert badges on comments when Q&A is enabled

CORL-856

* Update mutation responses and tests due to added expert fields

CORL-856

* Use EXPERT user tags to denote expert users

Removes the need for viewerIsExpert and
authorIsExpert loader/resolvers on Stories
and Comments respectively.

CORL-856

* Show an unanswered comment stream when Q&A is enabled

CORL-879

* Do not visually show the unanswered tag

CORL-879

nick-funk committed:
[CORL-859] Convert Featured stream into Answered for Q&A (#2842)

* Create preliminary schema changes for Q&A

Adds a mode and expert User onto the StorySettings.
Adds a mode selection drop down on a story's
Configure tab.

CORL-863

* Allow multiple experts, remove form elements from search

Makes the previous expert user on a Q&A story now
an array of users who can be assigned.

Converts the previous form based search that was
pulled from the admin community area into a set
of events built on callbacks.

CORL-863

* Create addExpertToStory mutation

CORL-863

* Create removeExpertFromStory mutation

CORL-863

* Conditionally show the the expert selection options

CORL-863

* Create a dropdown search control for Q&A experts

CORL-863

* Fixing up tests to match new QA stream options

Adds a few localization fixes to make sure tests
pass.
Updates existing snapshots.

CORL-863

* Add load more button to expert search list

CORL-863

* Update experts query to match react upgrades

CORL-863

* Move the Q&A config to its own section under stream config

Create enable and disable Q&A mutations/button toggle.

CORL-863

* Fix alignment and layout of expert list items

CORL-863

* Define translations and update tests

CORL-863

* Use official copy for Q&A config

CORL-863

* Show expert badges on comments when Q&A is enabled

CORL-856

* Update mutation responses and tests due to added expert fields

CORL-856

* Use EXPERT user tags to denote expert users

Removes the need for viewerIsExpert and
authorIsExpert loader/resolvers on Stories
and Comments respectively.

CORL-856

* Show an unanswered comment stream when Q&A is enabled

CORL-879

* Create preliminary schema changes for Q&A

Adds a mode and expert User onto the StorySettings.
Adds a mode selection drop down on a story's
Configure tab.

CORL-863

* Do not visually show the unanswered tag

CORL-879

* Allow multiple experts, remove form elements from search

Makes the previous expert user on a Q&A story now
an array of users who can be assigned.

Converts the previous form based search that was
pulled from the admin community area into a set
of events built on callbacks.

CORL-863

* Create addExpertToStory mutation

CORL-863

* Create removeExpertFromStory mutation

CORL-863

* Create a dropdown search control for Q&A experts

CORL-863

* Fixing up tests to match new QA stream options

Adds a few localization fixes to make sure tests
pass.
Updates existing snapshots.

CORL-863

* Add load more button to expert search list

CORL-863

* Update experts query to match react upgrades

CORL-863

* Move the Q&A config to its own section under stream config

Create enable and disable Q&A mutations/button toggle.

CORL-863

* Fix alignment and layout of expert list items

CORL-863

* Define translations and update tests

CORL-863

* Show expert badges on comments when Q&A is enabled

CORL-856

* Use official copy for Q&A config

CORL-863

* Update mutation responses and tests due to added expert fields

CORL-856

* Use EXPERT user tags to denote expert users

Removes the need for viewerIsExpert and
authorIsExpert loader/resolvers on Stories
and Comments respectively.

CORL-856

* Create the answered stream for Q&A

CORL-859

* Sort the Q&A on Most Voted by default

CORL-859

* Fix type mismatch between post comment form and fragments

CORL-859

tessalt commented on src/core/client/stream/tabs/Comments/Comment/CommentContainer.tsx:
```diff
@@ -210,6 +224,16 @@ export class CommentContainer extends Component<Props, State> {
+              <Icon size="xs" className={styles.tagIcon}>
+                check
+              </Icon>
+              ANSWERED
```
are we inside a localization tag here?

tessalt commented on src/core/client/stream/tabs/Comments/Comment/ReactionButton/ReactionButtonContainer.tsx:
```diff
@@ -80,8 +81,8 @@ class ReactionButtonContainer extends React.Component<Props> {
         reacted={reacted}
-        label={label}
-        labelActive={labelActive}
+        label={this.props.isQA ? "Vote" : label}
```
should "vote" and "voted" be localized?

tessalt commented on src/core/client/stream/tabs/Comments/Comment/UserTagsContainer.tsx:
```diff
@@ -1,26 +1,49 @@
+            <Icon size="xs" className={styles.icon}>
+              star
+            </Icon>
+            EXPERT
```
missing Localization tag

tessalt commented on src/core/client/stream/tabs/Comments/Stream/AnsweredCommentsTab/AnsweredCommentContainer.css:
```diff
@@ -0,0 +1,29 @@
+}
+
+.gotoArrow {
+  display: inline-block;
```
`inline-block` often indicates we could be using flexbox.

tessalt commented on src/core/client/stream/tabs/Comments/Stream/AnsweredCommentsTab/AnsweredCommentContainer.tsx:
```diff
@@ -0,0 +1,213 @@
+              viewer={viewer}
+            />
+          )}
+          <Box ml={1} container="span">
```
lot of uses of `mt` props in this component that could maybe be simplified with `HorizontalGutter` but no promises.

nick-funk replied:
I'll be honest, this was copy pasted from the original Featured comment stream. Going to prioritize the other more serious comments on the PR, but I'll see if this can be fixed. 👍

tessalt commented on src/core/client/stream/tabs/Comments/Stream/AnsweredCommentsTab/AnsweredCommentsContainer.tsx:
```diff
@@ -0,0 +1,182 @@
+  }, [loadMore, beginLoadMoreEvent, props.story.id]);
+  const comments = props.story.featuredComments.edges.map(edge => edge.node);
+  return (
+    <>
```
i don't think this fragment wrapper is needed.

tessalt commented on src/core/client/stream/tabs/Comments/Stream/UnansweredCommentsTab/UnansweredCommentsTabContainer.tsx:
```diff
@@ -0,0 +1,329 @@
+    (props.story.comments.viewNewEdges &&
+      props.story.comments.viewNewEdges.length) ||
+    0;
+  return (
```
i think this all makes sense but it is a lot

tessalt replied:
this is a useless comment lol ignore.

nick-funk replied:
Eh, I agree, will do what I can. Was original copy pasted from Featured stream, but that doesn't mean we can't clean it up! 😄

tessalt commented on src/core/client/stream/tabs/Configure/ConfigureStream/ConfigureStreamContainer.tsx:
```diff
@@ -23,6 +23,8 @@ class ConfigureStreamContainer extends React.Component<Props> {
     data: UpdateStorySettingsInput["settings"],
     form: FormApi
   ) => {
+    window.console.log(data);
```
🚒

tessalt commented on src/core/client/stream/tabs/Configure/Q&A/ExpertListItem.css:
```diff
@@ -0,0 +1,126 @@
+
+@media (min-width: 550px) {
+  .usernameEmail {
+    display: inline-block;
```
again this miiiight make more sense as a flexbox thing.

tessalt commented on src/core/client/stream/tabs/Configure/Q&A/ExpertListItem.css:
```diff
@@ -0,0 +1,126 @@
+  letter-spacing: 0.0125em;
+}
+
+.username {
```
did these few rules get repeated a few times here?

tessalt commented on src/core/client/stream/tabs/Configure/Q&A/ExpertListItem.tsx:
```diff
@@ -0,0 +1,49 @@
+        {username && <span className={styles.username}>{username}</span>}
+        {email && (
+          <span className={styles.email}>
+            {"("}
```
can we put this all one one line somehow?

tessalt commented on src/core/client/stream/tabs/Configure/Q&A/ExpertSearchItem.css:
```diff
@@ -0,0 +1,28 @@
+}
+
+.username {
+  display: inline-block;
```
is this the same styling as was used in the List Item?

tessalt commented on src/core/client/stream/tabs/Configure/Q&A/ExpertSelectionContainer.tsx:
```diff
@@ -0,0 +1,353 @@
+
+  const clearSearchFilter = useCallback(() => {
+    const searchInputNode: any = searchFieldRef.current;
+    searchInputNode.value = "";
```
can we manage the field state in a more react-y way?

nick-funk replied:
I thought ref's were part of React? 😆 
I get what you mean though, I'll try and find a cleaner way to clear this field.

tessalt commented on src/core/server/services/comments/pipeline/phases/expert.ts:
```diff
@@ -0,0 +1,31 @@
+  GQLTAG,
+} from "coral-server/graph/schema/__generated__/types";
+
+export const expert: IntermediateModerationPhase = ({
```
can we add a comment here or change the function name? "phases/expert" sounds ambiguous/vague imo.

tessalt reviewed (dismissed):
missing some localization tags + a couple UI questions & nitpicks.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -2590,6 +2619,20 @@ type StorySettings {
+  This will determine the appearance of the stream and how it functions.
+  This is an optional parameter and if unset, defaults to commenting.
+  """
+  mode: STORY_MODE
```
Instead of the documentation describing the default, apply the default in the resolver.

```suggestion
  mode: STORY_MODE!
```

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -5847,13 +6003,45 @@ type Mutation {
+  disableQAOnStory will disable Q&A mode on a story.
+  """
+  disableQAOnStory(input: DisableQAInput!): DisableQAPayload!
+    @auth(roles: [ADMIN, MODERATOR])
```
Instead of discrete `enable/disableQA` mutations, instead use `updateStoryMode`.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -5847,13 +6003,45 @@ type Mutation {
+  """
+  removeExpertFromStory(
+    input: RemoveExpertInput!
+  ): RemoveExpertPayload! @auth(roles: [ADMIN, MODERATOR])
```
Instead of using `from/to` in mutation, prefer `addStoryExpert` or `removeStoryExpert`.

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -929,7 +929,7 @@ export async function retrieveStoryCommentTagCounts(
     // `"tags.type": { $exists: true }` to ensure that we are using the
     // specified index.
-    "tags.type": GQLTAG.FEATURED,
+    "tags.type": { $exists: true },
```
Adjust the documentation above this line.

wyattjoh commented on src/core/server/models/story/index.ts:
```diff
@@ -501,3 +521,137 @@ export async function updateStoryLastCommentedAt(
+    story.settings.expertIDs.some(e => e === userID)
+  ) {
+    return story;
+  }
```
This check should be completed in the services layer, and not here.

nick-funk replied:
I disagree, this is important to ensure that we don't add a duplicate expert ID onto the story.

I don't want safety checks spread across a service layer as well as the model (repository). This is a data specific problem, the service does not care about duplicates, it cares about the adding/removing. These checks I believe belong in the data layer.

wyattjoh replied:
So, given that the below database code is using a `findOneAndUpdate` method, it actually performs the find operation for you, for free without any additional database access being required. The use of the `$addToSet` operation also actually prevents duplicates from being added to the set, thereby ensuring that the userID is not duplicated.

I would check to see the remainder of the function additions in this file do not query the database unnecessarily for these operations. You can instead just do:

```ts
if (!result.value) {
  throw new StoryNotFoundError(storyID);
}
```

Which can accomplish the same check without any additional database access.

The idea/drive behind this suggestion is that we want to ensure that we use atomic update operations when making document updates because we can then remove the pre-checks before these updates are performed.

I would also ensure that new functions added to this file include the name `Story` in the function name to stay consistent with the existing function naming.

wyattjoh commented on src/core/server/models/story/index.ts:
```diff
@@ -501,3 +521,137 @@ export async function updateStoryLastCommentedAt(
+    !story.settings.expertIDs.some(e => e === userID)
+  ) {
+    return story;
+  }
```
Same as before, should be moved to services layer.

nick-funk replied:
Disagree as per previous comment.

wyattjoh commented on src/core/server/services/comments/pipeline/phases/approve.ts:
```diff
@@ -0,0 +1,41 @@
+    result.tags.push({
+      type: GQLTAG.FEATURED,
+      createdAt: now,
+    });
```
This tag assignment should be moved to the `expert.ts` phase, which would simplify this phase to instead return a status `APPROVED` instead of editing the shared `result` from above.

wyattjoh commented on src/core/server/services/comments/pipeline/phases/unanswered.ts:
```diff
@@ -0,0 +1,52 @@
+
+  // If we have experts and the user is not one
+  // of them, this is an unanswered question.
+  if (!story.settings.expertIDs.some(id => id === comment.authorID)) {
```
```suggestion
  if (story.settings.expertIDs.every(id => id !== author.id)) {
```

wyattjoh replied:
Also, because the above logical test and this one both are making a decision based on the same domain (both are checking to see if this user is not an expert), it makes sense to me to combine these into the same logic test. It would instead read:

> If we have no experts, or the current author is not an expert, than this is an `UNANSWERED` comment.

wyattjoh commented on src/core/server/services/comments/pipeline/phases/expert.ts:
```diff
@@ -0,0 +1,31 @@
+  story,
+}): IntermediatePhaseResult | void => {
+  if (
+    author &&
```
This check isn't required.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -4242,6 +4286,8 @@ input UpdateStorySettings {
   """
   messageBox: StoryMessageBoxInput
+
+  mode: STORY_MODE
```
Because the story mode changes are done in another mutation, this should be removed.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -2653,6 +2696,7 @@ type Story {
     first: Int = 20 @constraint(max: 50)
     orderBy: COMMENT_SORT = CREATED_AT_DESC
     after: Cursor
+    tag: TAG
```
I would add a line in the docs that we should deprecate this `tag` in the future if we go deeper down Q&A by way of adding a `unansweredComments` edge to the story as we did for `featuredComments`.

wyattjoh commented on src/core/server/graph/loaders/Comments.ts:
```diff
@@ -200,12 +200,20 @@ export default (ctx: Context) => ({
+          ? {
+              $elemMatch: { type: tag },
+            }
+          : undefined,
```
Instead prefer:

```ts
...tagFilter(tag),
```

As seen: https://github.com/coralproject/talk/blob/29742efb4b516ca9c6d9a5ada6cc2279a282c4c8/src/core/server/graph/loaders/Comments.ts#L144

wyattjoh commented on src/core/client/stream/tabs/Configure/Q&A/QAConfigContainer.tsx:
```diff
@@ -0,0 +1,58 @@
+    }
+  }, [waiting, setWaiting, story, disableQA, enableQA]);
+
+  return story.settings.mode === "QA" ? (
```
Should use enum constants instead of inlined strings.

wyattjoh commented on src/core/client/stream/tabs/Configure/Q&A/QAConfigContainer.tsx:
```diff
@@ -0,0 +1,58 @@
+  const handleOnClick = useCallback(async () => {
+    if (!waiting) {
+      setWaiting(true);
+      if (story.settings.mode === "COMMENTS") {
```
Should use enum constants instead of inlined strings.

wyattjoh commented on src/core/client/stream/tabs/Configure/Q&A/AddExpertMutation.ts:
```diff
@@ -0,0 +1,51 @@
+    environment: Environment,
+    input: MutationInput<AddExpertMutation> & {
+      username: string;
+      email: string;
```
Why does this mutation accept a username/email address?

wyattjoh commented on src/core/client/stream/tabs/Comments/Stream/StreamContainer.tsx:
```diff
@@ -240,12 +291,29 @@ export const StreamContainer: FunctionComponent<Props> = props => {
+              </TabPane>
+            )}
+            {props.story.settings.mode === GQLSTORY_MODE.QA && (
+              <TabPane
+                className={CLASSES.allCommentsTabPane.$root}
+                tabID="UNANSWERED_COMMENTS"
+              >
+                <UnansweredCommentsTab />
+              </TabPane>
```
The same logic test is performed above checking for Q&A mode, perhaps moving this into a fragment in the previous section?

wyattjoh commented on src/core/client/stream/tabs/Comments/Stream/StreamContainer.tsx:
```diff
@@ -139,6 +142,12 @@ export const StreamContainer: FunctionComponent<Props> = props => {
+
+      // If we are in Q&A mode, we default to most voted
+      // sorting by default
+      if (props.story.settings.mode === GQLSTORY_MODE.QA) {
+        setLocal({ commentsOrderBy: "REACTION_DESC" });
+      }
```
These should be included in the `useEffect` dependencies, otherwise updates to the story will not cause this to be hit again.

wyattjoh commented on src/core/client/stream/tabs/Comments/Stream/AllCommentsTab/AllCommentsTabContainer.tsx:
```diff
@@ -153,18 +159,38 @@ export const AllCommentsTabContainer: FunctionComponent<Props> = props => {
+                There are no questions yet. Why don't you ask one?
+              </CallOut>
+            </Localized>
+          )}
```
Given the deep nesting of related checks, it makes more sense to me to nest these into an external component:

```tsx
const NoQuestions = ({ mode, isClosed }) => {
  if (mode === GQLSTORY_MODE.QA) {
    if (isClosed) {
      return <Localized /* ... */ />;
    } else {
      return <Localized /* ... */ />;
    }
  } else if (mode === GQLSTORY_MODE.COMMENTS) {
    if (isClosed) {
      return <Localized /* ... */ />;
    } else {
      return <Localized /* ... */ />;
    }
  }

  return null;
};

// ...

{comments.length <= 0 && <NoComments mode={props.story.settings.mode} isClosed={props.story.isClosed} />}
```

nick-funk committed:
Adding localizations for Q&A tags

nick-funk committed:
Hide feature flags in Q&A streams

nick-funk committed:
Allow experts to clear answered questions

Can click a button that shows up in the unanswered
tab under answered questions to refresh the stream,
clearing the answered questions from that tab.

nick-funk committed:
Show arrow upvote icon when in Q&A mode

Also localized the upvote text so it can
be translated.

nick-funk committed:
Hide mod/report buttons on answered questions

nick-funk committed:
Remove unnecessary fragment container

nick-funk committed:
Remove errant debug console log

nick-funk committed:
Make story mode required on story settings

nick-funk committed:
Make remove button outlined, not filled

nick-funk committed:
Further schema changes around Q&A and experts

Rename add/remove story expert to removeStoryExpert
and addStoryExpert naming.

Replace enableQA and disableQA mutators with single
updateStoryMode mutator.

nick-funk committed:
Remove story mode from UpdateStorySettings

nick-funk committed:
Replace inline string val's with enum

nick-funk committed:
add dependencies to useEffect

nick-funk committed:
docs cleanup around tags.type index check

nick-funk committed:
Approve a question when it is answered in Q&A

Approves using the author's id as the moderator

nick-funk committed:
Add comment around use of TAG on comments edge

nick-funk committed:
Use tagFilter instead of $elemMatch to filter by tags

nick-funk committed:
Improve responsive styles for expert list items

nick-funk committed:
Update copy to "Done" from "Remove answered questions"

nick-funk committed:
Text styling for no users found text

Styles the expert search list to have proper
text styling when no users are found for the
search keyword.

nick-funk committed:
Remove duplicate checks around story experts

Not necessary as Mongo does this for us.

nick-funk committed:
Fix a missed "sort imports" during rebase

nick-funk committed:
Refactor Q&A moderation phases for clarity

Simplify logic and update comments.
Rename: "answered" -> "tagExpertAnswers"
Rename: "unanswered" -> "tagUnansweredQuestions"

nick-funk committed:
Remove username & email from add expert mutation

nick-funk committed:
Format expert list emails with Localized

nick-funk committed:
Break out no comments logic into fragment

nick-funk committed:
Remove ref handling from expert search field

Use value assignment on TextField instead.

nick-funk committed:
Replace Box with Flex and CSS

nick-funk committed:
Show Q&A tooltip on Answered tab

kgardnr committed:
Merge branch 'master' into feature/qa

kgardnr committed:
Merge branch 'master' into feature/qa
